### PR TITLE
refactor(linux): Refactor processing actions in separate methods 🚧

### PR DESCRIPTION
**Pull request by ermshiperete**
## User Testing

@keymanapp/testers 

- [x] Please perform a basic acceptance test on Keyman for Linux. There should be no visible change from previous stable-14.0 release.

See [wiki](https://github.com/keymanapp/keyman/wiki/How-to-test-artifacts-from-pull-requests-for-Keyman-for-Linux) for instructions how to get the artifacts for Linux. The artifacts are available [here](https://jenkins.lsdev.sil.org/view/Keyman/view/Pipeline/job/pipeline-keyman-packaging/view/change-requests/job/PR-5452/lastSuccessfulBuild/artifact/).

### CODE REVIEW

File: linux/ibus-keyman/src/engine.c [hunk 1]
```diff
@@ -459,87 +459,84 @@ ibus_keyman_engine_destroy (IBusKeymanEngine *keyman)
...
+    KM_KBP_VKEY_6,       //        public static final int KEY_6 = 0x07;
+    KM_KBP_VKEY_7,       //        public static final int KEY_7 = 0x08;
+    KM_KBP_VKEY_8,       //        public static final int KEY_8 = 0x09;
+    KM_KBP_VKEY_9,       //        public static final int KEY_9 = 0x0A;
+    KM_KBP_VKEY_0,       //        public static final int KEY_0 = 0x0B;
+    KM_KBP_VKEY_HYPHEN,  //        public static final int KEY_MINUS = 0x0C;
+    KM_KBP_VKEY_EQUAL,   //        public static final int KEY_EQUALS = 0x0D;
+    KM_KBP_VKEY_BKSP,    //        public static final int KEY_BACKSPACE = 0x0E;
+    KM_KBP_VKEY_TAB,     //        public static final int KEY_TAB = 0x0F;
+    KM_KBP_VKEY_Q,       //        public static final int V_R = 0x13;
```
Comment (darcywong00) [inline linux/ibus-keyman/src/engine.c:500]: Did you intend to remove keys <kbd>W</kbd>, <kbd>E</kbd>, and <kbd>R</kbd>?

Comment (ermshiperete) [reply linux/ibus-keyman/src/engine.c:500]: ehhh - no. But who needs <kbd>W</kbd>, <kbd>E</kbd>, and <kbd>R</kbd> anyways :smile: 

Thanks!

**Comment by ermshiperete** [comment]
## User Testing

@keymanapp/testers 

1. Please perform a basic acceptance test on Keyman for Linux. There should be no visible change from previous stable-14.0 release.

See [wiki](https://github.com/keymanapp/keyman/wiki/How-to-test-artifacts-from-pull-requests-for-Keyman-for-Linux) for instructions how to get the artifacts for Linux. The artifacts are available [here](https://jenkins.lsdev.sil.org/view/Keyman/view/Pipeline/job/pipeline-keyman-packaging/view/change-requests/job/PR-5452/lastSuccessfulBuild/artifact/).

**Review by mcdurdin — APPROVED** [review]
Changes LGTM. I see varying brace styles in the refactored code -- in some places you've changed to put opening braces onto new lines, and in other places you haven't. Can you make the new code consistent with the style guide?

**Comment by ermshiperete** [comment]
> Changes LGTM. I see varying brace styles in the refactored code -- in some places you've changed to put opening braces onto new lines, and in other places you haven't. Can you make the new code consistent with the style guide?

done

**Comment by MakaraSok** [comment]
```
Access Denied
makara is missing the Overall/Read permission
```

Please give me permission on Jenkins to access the artifact.

**Comment by MakaraSok** [comment]
The artifacts were downloaded from [here](https://jenkins.lsdev.sil.org/view/Keyman/view/Pipeline/job/pipeline-keyman-packaging/view/change-requests/job/PR-5452/10/artifact/).

1. The `archive.zip` was downloaded from the link above and unzipped to `Downloads` folder
2. From the terminal, cd to the location of the unzipped folder (`~/Downloads)
3. executed this command
```
sudo dpkg -i \
{ibus-keyman,ibus-kmfl,keyman,keyman-config,keyman-keyboardprocessor,kmflcomp,libkmfl,libkmnkbp0,python3-keyman-config}*$(lsb_release -s -c)*.deb

```
The result is an error: `dpkg: error: cannot access archive 'keyman-config*focal*.deb': No such file or directory`. 
Here is the screenshot:

image

**Comment by ermshiperete** [comment]
> ```
> sudo dpkg -i \
> {ibus-keyman,ibus-kmfl,keyman,keyman-config,keyman-keyboardprocessor,kmflcomp,libkmfl,libkmnkbp0,python3-keyman-config}*$(lsb_release -s -c)*.deb
> ```
> 
> The result is an error: `dpkg: error: cannot access archive 'keyman-config*focal*.deb': No such file or directory`.

Oops, `keyman-config` isn't supposed to be in that list. Sorry about that.

**Comment by MakaraSok** [comment]
This occurs after running the command:

```
dpkg: error processing package python3-keyman-config (--install):
 dependency problems - leaving unconfigured
Setting up ibus-keyman (15.0.86-1+focal1) ...
dpkg: dependency problems prevent configuration of keyman:
 keyman depends on python3-keyman-config; however:
  Package python3-keyman-config is not configured yet.

dpkg: error processing package keyman (--install):
 dependency problems - leaving unconfigured
Setting up kmflcomp (15.0.86-1+focal1) ...
Setting up libkmfl0:amd64 (15.0.86-1+focal1) ...
Setting up libkmfl-dev:amd64 (15.0.86-1+focal1) ...
Setting up ibus-kmfl (15.0.86-1+focal1) ...
Processing triggers for gnome-menus (3.36.0-1ubuntu1) ...
Processing triggers for desktop-file-utils (0.24-1ubuntu3) ...
Processing triggers for mime-support (3.64ubuntu1) ...
Processing triggers for libglib2.0-0:amd64 (2.64.6-1~ubuntu20.04.3) ...
Processing triggers for hicolor-icon-theme (0.17-2) ...
Processing triggers for man-db (2.9.1-1) ...
Processing triggers for shared-mime-info (1.15-1) ...
Processing triggers for libc-bin (2.31-0ubuntu9.2) ...
Errors were encountered while processing:
 python3-keyman-config
 keyman

```

**Comment by MakaraSok** [comment]
# Keyman for Linux Acceptance Test Procedures #

* **_No note is given under the checkbox when a test behaves as expected._**
* **_If FAILED or noteworthy, the result and comment are given after `>` in the line_**
* **_Empty checkboxes are not applicable._**


## Visual Inspection of master/beta branches
- [x] Build Status/statusIcon) Verify latest CI build of [master](https://build.palaso.org/viewType.html?buildTypeId=KeymanLinux_Master) is successful
- [ ] Build Status/statusIcon) Verify latest CI build of [beta](https://build.palaso.org/viewType.html?buildTypeId=KeymanLinux_Beta) is successful
- [ ] Verify `/HISTORY.md` contains all the current changes

## Gather Assets for Testing ##
- [ ] Ubuntu 18.04 Bionic or
- [x] Ubuntu 20.04 Focal or
- [ ] Ubuntu 20.10 Groovy - VMN or physical okay

## Setup Steps ##
- [x] Uninstall previous versions of Keyman for Linux on the system

# Keyman for Linux #

## Basic installation ##
- [x] Add the ppa
```
sudo add-apt-repository ppa:keymanapp/keyman-beta
sudo apt update
```
- [x] Install keyman and onboard
```
sudo apt install keyman onboard-keyman
```
- [x] From the launcher, start "Keyman Keyboard"
> **FAILED:** - Keyman does not start from launcher.
- [x] Verify "Keyman Configuration" panel appears
> **FAILED:** - It only opens when running `km-config` in the terminal
- [x] Click "Download", enter a keyboard to search for, and click the keyboard
- [x] Click "Install keyboard"
- [x] If the package includes a readme file, verify that it appears in the install window
- [x] If the package includes a welcome file, verify that it appears after installation
- [x] Verify the new keyboard appears in "Keyman Configuration"
- [x] Verify the keyboard was added to GNOME
    - [x] From Ubuntu, start "Settings" --> "Region & Language"
    - [x] Verify the keyboard appears in the "Input Sources" list
    - [x] Verify the language tag for the current keyboard appears on the top right of the screen (`"en"` for English")
          (if the keyboard indicator doesn't show, reboot)
- [x] Click Windows-spacebar to switch to the newly installed keyboard
> **PASSED** - when press and hold the Windows key and then press and release the spacebar, the desired keyboard is highlighted on the screen. Release the spacebar to select a keyboard.
- [x] In a text editor (gedit), verify that the newly installed keyboard can type

## Add a keyboard for an additional language

- [x] From the launcher, start "Keyman Keyboard"
> **FAILED:** - Keyman does not start from launcher.
- [x] Verify "Keyman Configuration" panel appears
> **FAILED:** - It only opens when running `km-config` in the terminal
- [x] Click "Download", search for "icelandic", and click the "EuroLatin (SIL) (Icelandic language)" keyboard
- [x] Click "Install keyboard"
- [x] Verify that the readme appears in the install window
> **FAILED:** - It doesn't appear by default. The details appear by default. To show readme, click on `README` tab.
- [x] Click "Install" button
- [x] Verify that the welcome file appears after installation
- [x] Verify the new keyboard appears in "Keyman Configuration"
- [x] Verify the new keyboard appears on the top right of the screen (similar to "Icelandic (EuroLatin (SIL)) - is")
- [x] Add the keyboard for an additional language in GNOME
    - [x] From Ubuntu, start "Settings" --> "Region & Language"
    - [x] In "Input Sources", click "+" to add input source
    - [x] Search for the language (e.g. German) and select the "EuroLatin (SIL)" keyboard for that language (e.g. "German (EuroLatin (SIL))")
> **PASSED** - but a few more steps are to take to be able to search for "German," i.e. click on three little vertical dots > type in the search term > click on "Other" and then find "German (EuroLatin (SIL))" and click on it to enable the green "Add" button on the top right of the window. 
image

    - [x] Click "Add"
    - [x] Verify the language/keyboard combination appears in the "Input Sources" list
    - [x] Verify the language tag for the current keyboard appears on the top right of the screen (`"en"` for English")
- [x] Click Windows-spacebar to switch to the newly installed keyboard
- [x] In a text editor (gedit), verify that the newly installed keyboard can type (e.g. typing `:a` should result in `ä`)

## Install a package from disk
- [x] From the "Keyman Configuration" panel, click "Install"
- [x] Browse to a local .kmp file on disk and click "Open" to install it
- [x] If the package includes a readme file, verify that it appears in the install window
> **PASSED** - but it's not readme that appears by default. Details appear by default.
> Cameroon QWERTY keyboard contains a number of language, but when installed this way the default keyboard (Afade) was installed. It might be helpful to have a step in between where the user can choose a language to install the keyboard with. The mobile versions of Keyman do this.
- [x] If the package includes a welcome file, verify that it appears after installation
- [x] Verify the keyboard was added to GNOME
    - [x] From Ubuntu, start "Settings" --> "Region & Language"
    - [x] Verify the keyboard appears in the "Input Sources" list
    - [x] Verify the language tag for the current keyboard appears on the top right of the screen
- [x] Click Windows-spacebar to switch to the newly installed keyboard
- [x] In a text editor, verify that the newly installed keyboard can type

## About keyboard
- [x] From the "Keyman Configuration" panel, select a keyboard and click "About"
- [x] Verify package and keyboard information is displayed
- [x] For a keyboard downloaded from keyman.com, verify a QR Code link is displayed

## Keyboard help
- [x] From the "Keyman Configuration" panel, select a keyboard and click "Help"
- [ ] If the package does not include a welcome file verify that the "Help" button is disabled
- [x] If the package includes a welcome file, verify that it appears.

## Options
- [x] From the "Keyman Configuration" panel, select the "EuroLatin (SIL)" keyboard and click "Options"
- [x] Verify the Options dialog appears.
- [x] Select a different keyboard that doesn't provide options
- [x] Verify the Options button is disabled.

## On Screen Keyboard tests
- [x] From the launcher, start "onboard"
- [x] Verify the OSK for the current keyboard appears at the bottom of the screen
- [x] Verify that the default OSK (en-US) is shown if the current keyboard doesn't have one
- [x] Click Windows-spacebar to switch to another keyboard than has an OSK
- [x] Verify the OSK switches to match the current keyboard
- [x] In a text editor, verify clicking on an OSK key produces expected output

## On Screen Keyboard configuration
- [x] From the launcher, start "onboard settings"
- [x] Verify the "Onboard Preferences" panel appears

## Uninstallation
- [x] Remove the keyman package
```
sudo apt --autoremove remove keyman
```
- [x] Verify no errors are displayed
- [x] Verify the directory and files in `/usr/lib/python3/dist-packages/keyman_config` are gone
> **FAILED:** - `keyman_config` is still in the directory.
image

## What's New Tests
- [ ] Refer to the new changes in `linux/history.md` and verify functionality

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.87-alpha](https://keyman.com/downloads/releases/alpha/15.0.87)

**Pull request merged**